server/src/main/java/org/opensearch/search/SearchService.java

```diff
@@ -1114,6 +1141,55 @@ private DefaultSearchContext createSearchContext(ReaderContext reader, ShardSear
         return searchContext;
     }
 
+    private void writeTaskResourceUsage(SearchShardTask task) {
+        try {
+            // Get resource usages when task starts
+            ThreadResourceInfo threadResourceInfo = task.getActiveThreadResourceInfo(
+                Thread.currentThread().getId(),
+                ResourceStatsType.WORKER_STATS
+            );
+            if (threadResourceInfo == null) {
+                return;
+            }
+            Map<ResourceStats, ResourceUsageInfo.ResourceStatsInfo> startValues = threadResourceInfo.getResourceUsageInfo().getStatsInfo();
+            if (!(startValues.containsKey(ResourceStats.CPU) && startValues.containsKey(ResourceStats.MEMORY))) {
+                return;
+            }
+            // Get current resource usages
+            ResourceUsageMetric[] endValues = taskResourceTrackingService.getResourceUsageMetricsForThread(Thread.currentThread().getId());
+            long cpu = -1, mem = -1;
+            for (ResourceUsageMetric endValue : endValues) {
+                if (endValue.getStats() == ResourceStats.MEMORY) {
+                    mem = endValue.getValue();
+                } else if (endValue.getStats() == ResourceStats.CPU) {
+                    cpu = endValue.getValue();
+                }
+            }
+            if (cpu == -1 || mem == -1) {
+                logger.debug("Invalid resource usage value, cpu [{}], memory [{}]: ", cpu, mem);
+                return;
+            }
+
+            // Build task resource usage info
+            TaskResourceInfo taskResourceInfo = new TaskResourceInfo.Builder().setAction(task.getAction())
+                .setTaskId(task.getId())
+                .setParentTaskId(task.getParentTaskId().getId())
+                .setNodeId(clusterService.localNode().getId())
+                .setTaskResourceUsage(
+                    new TaskResourceUsage(
+                        cpu - startValues.get(ResourceStats.CPU).getStartValue(),
+                        mem - startValues.get(ResourceStats.MEMORY).getStartValue()
+                    )
+                )
+                .build();
+
+            threadPool.getThreadContext().removeResponseHeader(TASK_RESOURCE_USAGE);
```

Comment:
Should we ever have an existing response header here?

Reply:
Unfortunately yes. Say a search query has 2 phases: `query` and `fetch`. When the query phase finishes, it writes to the responseHeader. When the fetch phase finishes, the responseHeader will still contain the resource usages info of the query phase if we don't clear it. In that case, the data piggybacked during the fetch phase will contain both the query and fetch phases, so we will be double counting the query phase resource usages.

Another solution could be accumulate the results in the thread context and only send back in the phase at the very end. But that will make the logic less stateless and also it would be harder to handle the failed phases. Thus I think simply removing the previous resource usage in the header (if there's any) every time could be a more elegant solution here.

Reply:
Oh... got it. Yeah, given that setup, removing the existing header makes sense.

I've seen a few times (not in OpenSearch, but on other projects) where per-request state stored in a `ThreadLocal` leaks across requests (e.g. because the first request threw an exception, so state wasn't cleared). I just wanted to make sure we're not papering over those kinds of bugs.

Maybe add a comment highlighting that the existing header would have come from an earlier phase in the same request?

Reply:
Just updated!

Reply:
Does the addResponseHeaders logic not replace existing key value?